rfcs/0008-leaderless-dns.md

```diff
@@ -0,0 +1,221 @@
+
+# Summary
+[summary]: #summary
+
+Enable the MGC to manage DNS for one or more hostnames shared across one or more clusters, where a central "hub" cluster is not required or may not be available, and increase the robustness of the DNS Health checks by executing them from all the available clusters.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, the MGC has limited functionality for a single cluster, or for multiple clusters without OCM. Having the MGC enabled to manage DNS in these situations significantly eases onboarding users to the advantages of the MGC.
+
+# Diagrams
+[diagrams]: #diagrams
+
+- [Overview](./0008-leaderless-dns-assets/leaderless-dns-usecases.jpg)
+- [P2P Health Checks](./0008-leaderless-dns-assets/leaderless-dns-healthchecks.jpg)
+- [DNS Pruning](./0008-leaderless-dns-assets/leaderless-dns-pruning.jpg)
+- [DNS Controller](./0008-leaderless-dns-assets/leaderless-dns-dns-operator.jpg)
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Single Cluster
+In a single cluster scenario, create a DNS Policy that references a DNS provider. When a matching host on a gateway is found to be in use, the IP addresses of that gateway will be created as A records for that host in the Zone file following the configuration outlined in the DNS Policy.
+
+### Single Cluster Health Checks
+If health checks are defined in the DNS Policy, these will also be created and used to remove unhealthy IPs from the Zone file, should all health checks fail, then all A records will be published to avoid an NXDOMAIN response.
+
+## Multiple Cluster
+In a multiple cluster scenario, every cluster requires a DNS Policy configured with a provider secret that specifies a unique `clusterID`, they should all reference the same zone(s) in the same DNS Provider for the same host. In each cluster, when a matching host on a gateway is found to be in use, the IP addresses of that gateway will be appended to the zone file - in a non-destructive manner - allowing multiple clusters to all add their gateways IP addresses to the same zone file, without destroying each other's values.
+
+### Multiple Cluster Health Checks
+If health checks are defined in the DNS Policy, these will be created on all clusters, there are then 2 types of health checks:
+1) Local Health Checks - used to check this clusters IP addresses are responding for the expected host.
+2) Remote Health Checks - used to check that the other IP addresses in the zone file are responding for that zones host.
+
+When a local health check is failing, this IP will not be published to the zone.
+
+When a remote health check is failing a TXT record is added/amended in the zone to note the failure - should the percentage of failing clusters exceed the `MultiClusterFailureThreshold` (configured in the DNS Policy), that unhealthy IP is removed from the zone by any cluster with an unhealthy check on that IP.
```

Comment:
could we add an example of the txt record. It would be important not to have sensitive information in this record

Reply:
yes, we will hash the clusterIDs when using them in the Zone.

Reply:
Can we add an example of the TXT record?

Reply:
I have added it, it's in the Remote Health Check Failing section.